src/lib/services/user-service.ts

```diff
@@ -239,50 +240,53 @@ class UserService {
         { username, email, name, password, rootRole }: ICreateUser,
         auditUser: IAuditUser = SYSTEM_USER_AUDIT,
     ): Promise<IUserWithRootRole> {
-        if (!username && !email) {
-            throw new BadDataError('You must specify username or email');
-        }
+        try {
+            if (!username && !email) {
+                throw new BadDataError('You must specify username or email');
+            }
 
-        if (email) {
-            Joi.assert(
+            if (email) {
+                Joi.assert(
+                    email,
+                    Joi.string().email({ ignoreLength: true }),
+                    'Email',
+                );
+            }
+
+            const user = await this.store.insert({
+                username,
                 email,
-                Joi.string().email({ ignoreLength: true }),
-                'Email',
-            );
-        }
+                name,
+            });
 
-        const exists = await this.store.hasUser({ username, email });
-        if (exists) {
-            throw new BadDataError('User already exists');
-        }
+            await this.accessService.setUserRootRole(user.id, rootRole);
 
-        const user = await this.store.insert({
-            username,
-            email,
-            name,
-        });
+            if (password) {
+                const passwordHash = await bcrypt.hash(password, saltRounds);
+                await this.store.setPasswordHash(
+                    user.id,
+                    passwordHash,
+                    disallowNPreviousPasswords,
+                );
+            }
 
-        await this.accessService.setUserRootRole(user.id, rootRole);
+            const userCreated = await this.getUser(user.id);
 
-        if (password) {
-            const passwordHash = await bcrypt.hash(password, saltRounds);
-            await this.store.setPasswordHash(
-                user.id,
-                passwordHash,
-                disallowNPreviousPasswords,
+            await this.eventService.storeEvent(
+                new UserCreatedEvent({
+                    auditUser,
+                    userCreated,
+                }),
             );
-        }
 
-        const userCreated = await this.getUser(user.id);
-
-        await this.eventService.storeEvent(
-            new UserCreatedEvent({
-                auditUser,
-                userCreated,
-            }),
-        );
+            return userCreated;
+        } catch (error) {
+            if (error.code === UNIQUE_CONSTRAINT_VIOLATION) {
```

Comment:
This is how to handle a concurrent problem and replace the previous validation, `hasUser` method. There are several patterns like this on the code base hence I follow them.